app/redux/redux-actions/adminActions.js

```diff
@@ -89,6 +91,20 @@ export const loadSettings = (data = {}) => {
     payload: data,
   };
 };
+export const restoreFormProgress = (data = {}) => {
+  return {
+    type: KEEP_FORM_CONTENT,
+    payload: data,
+  };
+};
+export const reduxKeepFormContent = ({ key, data, whole }) => {
+  const payload = { ...(whole || {}), [key]: data };
+  localStorage.setItem(ME_FORM_PROGRESS, JSON.stringify(payload));
```

Comment:
Ultimately, this is a very well thought out implementation; I'd go so far as to say a very clever idea, atta boy.  

I would _strongly recommend_ that you think about the following to make this more robust and extensible.

### Ideal Implementation
I'd like you to think about the concept of Drafts in applications like gmail.  That is essentially the same outcome we would want something like this to achieve.  Here are some general properties, I would expect it to have:
1. When i start filling out a form, i **won't** lose data any of the following scenarios:
- Navigate to a different part of the site and come back to this form
- Reload the page with the form
- Close the website, re-open and come back to this page
- When i Login with a different device
In all the above scenarios, I'd expect to be able to retrieve my drafts and continue where i left off. 

2. Proper access control is implemented
- No other application in the client can access the draft data except the creator.
- No other user can access the draft data.


Using local storage will introduce some limitations right away.  For example
1. In as much as we may be able to pull off 1(i), 1(ii), 1(iii)  with LocalStorage, we can't achieve the last property with it.
2. Secondly, we can't achieve 2 by ensuring proper and strict access control even though we could explore encrypting or encoding the data we store in the local storage as a stop-gap.

This means that, for the best implementation of this, we would like the backend to store this data.
For example, have a model called Draft that has fields, user, data (`a JSONField`), page_key, form_key, etc

With this, you can leverage the user's auth cookies and tokens to retrieve and save draft data back and forth.


## Non-Ideal Solution
Even though I will not recommend it, we could do the following:
1. Encrypt the data you are writing to the local storage with a key and a salt.
- This will ensure that no other application can make sense of it even if they read it.
- Same reason above for any user that chances on the data.  It is only the Massenergize application that can decode it.

The drawback here is: 
- Encryption should really never be done on the client side.  It means you expose your encryption algorithm and salt and key to the public
- It also means that, you can rotate your keys for encryption because, when you do, the previous key cant decode the old encrypted data, rendering the functionality moot.

Reply:
@Opoku-Agyemang - I have maintained that we don't need to preserve the draft content under these situations:

- Reload the page with the form
- Close the website, re-open and come back to this page
- When i Login with a different device

The primary use case where cadmins would expect to not lose draft content is if they accidently navigate to another part of the admin portal and come back and lose what they were typing.  (agree @kaatvds ?)

Other case where cadmins might be disappointed if they lose their content is:

1.  They get signed out of the admin portal after the time out.  Maybe they were half way through entering the content and decided to finish it the next day;

2. The site crashes.  Perhaps they hit submit and the form hangs.  This could be challenging, but perhaps the user sees the blank screen or error message, then hits "back" and "forward".  I guess this argues that the browser reload should restore content.  

3.  The browser closes and restarts (maybe because the computer restarts) and the browser opens up with the same tabs.  

This is pretty complicated, and all of this is pretty rare.  So I don't want to spend lots more time on it.   I do not think it is important to protect against the user signing on with a different device.